File: runtime/Jlib/buffer.java
package bigloo;

public class buffer
{
  public final byte[] buf;
  public int pos;

  public buffer( final int  size )
  {
    buf= new byte[size + 3];
    pos= 0;
  }

  public String toString()
  {
    return new String( buf, 0, pos );
  }

  public void add( final String  s )
  {
    add( s.getBytes() );
  }

  public void add( final byte[]  s )
  {
    if (pos+s.length > buf.length-3)
    {
      blit( "...".getBytes() );
      throw new RuntimeException( "end of buffer" );
    }
    blit( s );
  }

  public void add( final byte  b )
  {
    if (pos+1 > buf.length-3)
    {
      blit( "...".getBytes() );
      throw new RuntimeException( "end of buffer" );
    }
    buf[pos++]= b;
  }

  private void blit( final byte[]  s )
  {
    for ( int i= 0 ; i < s.length ; ++i )
      buf[pos++]= s[i];
  }
}

File: runtime/CSlib/buffer.cs
using System;
using System.Text;

namespace bigloo
{
  public sealed class buffer
  {
    public readonly byte[] buf;
    public int pos;

    public buffer( int  size )
    {
      buf= new byte[size + 3];
      pos= 0;
    }

    public override String ToString()
    {
      return foreign.newstring( buf, 0, pos );
    }

    public void add( String  s )
    {
      add( foreign.getbytes( s ) );
    }

    public void add( byte[]  s )
    {
      if (pos+s.Length > buf.Length-3)
      {
        blit( foreign.getbytes( "..." ) );
        throw new ApplicationException( "end of buffer" );
      }
      blit( s );
    }

    public void add( byte  b )
    {
      if (pos+1 > buf.Length-3)
      {
        blit( foreign.getbytes( "..." ) );
        throw new ApplicationException( "end of buffer" );
      }
      buf[pos++]= b;
    }

    private void blit( byte[]  s )
    {
      for ( int i= 0 ; i < s.Length ; ++i )
        buf[pos++]= s[i];
    }
  }
}
